oneflow/core/autograd/gradient_funcs/broadcast_binary_ops.cpp

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2020 The OneFlow Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include "oneflow/core/framework/op_expr_grad_function.h"
+#include "oneflow/core/framework/op_builder.h"
+#include "oneflow/core/framework/op_interpreter/op_interpreter_util.h"
+#include "oneflow/core/framework/op_expr.h"
+#include "oneflow/core/framework/op_expr_helper.h"
+
+namespace oneflow {
+namespace one {
+
+namespace {
+
+class ReduceSumLikeModule {
+ public:
+  ReduceSumLikeModule(const std::string& op_name) {
+    identity_op_ = op_expr_helper::IdentityOp(op_name).GetPtrOrThrow();
+    reshape_like_op_ = op_expr_helper::ReshapeLikeOp(op_name).GetPtrOrThrow();
+    reduce_sum_like_op_ = op_expr_helper::ReduceSumLikeOp({-1}, op_name).GetPtrOrThrow();
+  }
+
+  Maybe<Tensor> forward(const std::shared_ptr<Tensor>& input,
+                        const std::shared_ptr<Tensor>& like) const {
+    const auto& in_shape = *(input->shape());
+    const auto& like_shape = *(like->shape());
+    TensorTuple inputs{input};
+    AttrValueMap attrs;
+    std::shared_ptr<OpExpr> op = identity_op_;
+    if (in_shape != like_shape) {
+      const Shape& left_extended_shape =
+          CreateLeftExtendedShape(ShapeView(like_shape), in_shape.NumAxes());
+      if (in_shape == left_extended_shape) {
+        op = reshape_like_op_;
+      } else {
+        op = reduce_sum_like_op_;
+        const AxisVector& broadcast_axis_vec = left_extended_shape.Axes4BroadcastTo(in_shape);
+        JUST(attrs.SetAttr<std::vector<int32_t>>(
+            "axis", std::vector<int32_t>{broadcast_axis_vec.begin(), broadcast_axis_vec.end()}));
+      }
+      inputs.push_back(like);
+    }
+    return JUST(OpInterpUtil::Dispatch<Tensor>(*op, inputs, attrs));
+  }
+
+ private:
+  std::shared_ptr<OpExpr> identity_op_;
+  std::shared_ptr<OpExpr> reshape_like_op_;
+  std::shared_ptr<OpExpr> reduce_sum_like_op_;
+};
+
+}  // namespace
+
+class BroadcastBinary : public OpExprGradFunction<OpExprInterpState> {
+ public:
+  BroadcastBinary() = default;
+  virtual ~BroadcastBinary() = default;
+
+  virtual Maybe<void> Init(const OpExpr& op) {
+    const auto* fw_op_expr = dynamic_cast<const UserOpExpr*>(&op);
+    CHECK_NOTNULL_OR_RETURN(fw_op_expr);
+    op_name_ = fw_op_expr->op_name();
```

Comment:
```suggestion
    op_name_ = GradientOpName(fw_op_expr->op_name());
```

实现的grad function的op_name好像没有一个统一的规则

Reply:
后续有时间可以整理一下op_name这块，最终可以把名字带进FunctionNode中

Reply:
其实有一个规则的，通常就是在当前op name基础上加一个".grad"后缀，如果对应多个op，就把op name基础上再加一个当前op的一个描述。

Reply:
是，只是我看有的grad function加了，有的没有。如果统一有名字的话，op_name应该是父类的成员吧？

Reply:
都需要加的，可能有些地方遗漏了。这个op_name_是缓存的一个前向op name，不是后向的，后向name需要GradientOpName生成

Reply:
哦哦明白了